src/libs/elektra/hooks.c

```diff
@@ -0,0 +1,88 @@
+/**
+ * @file
+ *
+ * @brief
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#include <kdbinternal.h>
+
+void freeHooks (KDB * kdb, Key * errorKey)
+{
+	if (kdb->hooks.gopts.plugin != NULL)
+	{
+		elektraPluginClose (kdb->hooks.gopts.plugin, errorKey);
+	}
+}
+
+static size_t getFunction (Plugin * plugin, const char * functionName, Key * errorKey)
+{
+	size_t result = elektraPluginGetFunction (plugin, functionName);
+
+	if (result == 0)
+	{
+		ELEKTRA_ADD_INSTALLATION_WARNINGF (errorKey, "Plugin '%s' does not implement function '%s'", plugin->name, functionName);
+	}
+
+	return result;
+}
+
+static int initHooksGopts (KDB * kdb, Plugin * plugin, Key * errorKey)
+{
+	if (!plugin)
+	{
+		return -1;
+	}
+
+	kdb->hooks.gopts.plugin = plugin;
+
+	if ((kdb->hooks.gopts.kdbHookGoptsGet = (kdbHookGoptsGetPtr) getFunction (plugin, "hooks/gopts/get", errorKey)) == NULL)
+	{
+		return -1;
+	}
+
+	return 0;
+}
+
+static Plugin * loadPlugin (const char * pluginName, KeySet * modules, Key * errorKey)
+{
+	Key * openKey = keyDup (errorKey, KEY_CP_ALL);
+
+	Plugin * plugin = elektraPluginOpen (pluginName, modules, ksNew (0, KS_END), openKey);
+
+	if (!plugin)
+	{
+		ELEKTRA_ADD_INSTALLATION_WARNINGF (errorKey, "Could not load plugin '%s'", pluginName);
+		keyCopyAllMeta (errorKey, openKey);
+		keyDel (openKey);
+		return NULL;
+	}
+
+	return plugin;
+}
+
+static bool isGoptsEnabledByContract (const KeySet * contract)
+{
+	KeySet * dupContract = ksDup (contract); // We need to duplicate because contract is const, and ksLookupByName doesn't take const
```

Comment:
@markus2330 We should probably provide a variant of `ksLookup` that works with a `const KeySet *`.

AFAICT these are the reasons why we can't use `const`:
1. We may build the hashmap
2. `KDB_O_POP` could remove a key and `KDB_O_CREATE` could create a key
3. We may create a key from `meta:/default`
4. We update the internal cursor

AFAIK cases 2-4 will all be removed soon. IMO in cases like this one, a `const` variant could be useful to avoid the `ksDup`. The variant never builds the hashmap and just falls back to binary search, if the hashmap wasn't built yet.

Reply:
From the C Standard section 6.7.3:

> If an attempt is made to modify an object defined with a const-qualified type through use of an
lvalue with non-const-qualified type, the behavior is undefined

Seems pretty clear to me.

Since this is part of our public API we should really do things properly and not go against the standard without good reason. Here we don't know where the `KeySet` originally comes from, so casting a `const KeySet *` to a `KeySet *` is definitely not safe.

Reply:
I have created issue #4474 for this. Let's not discuss this any further in this PR.

Reply:
I noticed there is `ksDeepDup` in `ensureContract` (in `kdb.c`), so independent of the outcome of #4473, we can avoid this `ksDup`. We just need to pull the `ksDeepDup` (and the corresponding `ksDel`) out of `ensureContract` and into `kdbOpen`. Then we can move it before the `initHooks` call and just pass the deep-duped `KeySet`.